cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -272,58 +235,41 @@ private void initializeMessageIds() {
     });
   }
 
-  private void handleForbidden(PeerInfo peerInfo) throws IOException {
-    // Set tethering status to rejected.
-    store.updatePeerStatusAndTimestamp(peerInfo.getName(), TetheringStatus.REJECTED);
-  }
-
-  private void handleResponse(HttpResponse resp, PeerInfo peerInfo) throws IOException {
-    TetheringControlResponse[] responses;
-    try {
-      responses = GSON.fromJson(resp.getResponseBodyAsString(StandardCharsets.UTF_8), TetheringControlResponse[].class);
-    } catch (Exception e) {
-      LOG.error("Failed to parse response from peer {}: {}",
-                peerInfo.getName(),
-                resp.getResponseBodyAsString(StandardCharsets.UTF_8), e);
-      return;
-    }
-    if (peerInfo.getTetheringStatus() == TetheringStatus.PENDING) {
-      LOG.debug("Peer {} transitioned to ACCEPTED state", peerInfo.getName());
-      store.updatePeerStatusAndTimestamp(peerInfo.getName(), TetheringStatus.ACCEPTED);
-    } else {
-      // Update last connection timestamp.
-      store.updatePeerTimestamp(peerInfo.getName());
-    }
-    processTetheringControlResponse(responses, peerInfo);
-  }
-
-  private void processTetheringControlResponse(TetheringControlResponse[] responses, PeerInfo peerInfo)
-    throws IOException {
-    for (TetheringControlResponse response : responses) {
-      TetheringControlMessage controlMessage = response.getControlMessage();
-      switch (controlMessage.getType()) {
-        case KEEPALIVE:
-          LOG.trace("Got keepalive from {}", peerInfo.getName());
-          break;
-        case START_PROGRAM:
-          LOG.trace("Got start_program from {}", peerInfo.getName());
-          publishStartProgram(Bytes.toString(controlMessage.getPayload()), peerInfo);
-          break;
-        case STOP_PROGRAM:
-          LOG.trace("Got stop_program from {}", peerInfo.getName());
-          publishStopProgram(Bytes.toString(controlMessage.getPayload()));
-          break;
-        case KILL_PROGRAM:
-          LOG.trace("Got kill_program from {}", peerInfo.getName());
-          publishKillProgram(Bytes.toString(controlMessage.getPayload()));
-          break;
+  /**
+   * Processes control messages sent by the server and returns the id of the last processed message.
+   * Returns null if no messages were processed.
+   */
+  @Nullable
+  private String processTetheringControlResponse(TetheringControlResponseV2 response, PeerInfo peerInfo) {
+    String lastMessageId = null;
+    for (TetheringControlMessageWithId messageWithId : response.getControlMessages()) {
+      TetheringControlMessage controlMessage = messageWithId.getControlMessage();
+      try {
+        switch (controlMessage.getType()) {
+          case KEEPALIVE:
+            LOG.trace("Got keepalive from {}", peerInfo.getName());
```

Comment:
nit: these trace messages can be combined into a single one before the switch, with the type as a `{}` in the log message

Reply:
fixed